addons/addon-user-id-upgrade/packages/user-id-upgrade-post-deployment/lib/helpers/users.js

```diff
@@ -0,0 +1,146 @@
+    };
+    // A map between user object string and user id string, example {<json.stringify({ns, username})>: '<uid>'}
+    this.objIdMap = {
+      [systemObjId]: systemUid,
+    };
+    this.items = {}; // A list of user keyed by uid
+    this.usersTableName = usersTableName;
+    this.oldUsersTableName = oldUsersTableName;
+    this.report = report;
+  }
+
+  // Use this function to tell the UsersHolder about the rows (userItems) belonging to the new users table.
+  // This function, then, examines the row data and update its internal lookup map.
+  // userItems = an array of items. An item is the raw Item object returned by DynamoDB which is
+  // a map of AttributeValues. Example: item = { username: { 'S': 'root' }, ns: { 'S': 'internal' }, ... }
+  // For more information about what an AttributeValue is, see
+  // https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_AttributeValue.html
+  setItems(userItems) {
+    const tableName = this.usersTableName;
+    const tableReport = this.report.getTableReport(tableName);
+    _.forEach(userItems, item => {
+      const uid = parseAttributeValue(item.uid);
+      const username = parseAttributeValue(item.username);
+      const ns = parseAttributeValue(item.ns);
+
+      if (_.isUndefined(uid) || _.isUndefined(username) || _.isUndefined(ns)) {
+        const itemStr = JSON.stringify(item);
+        const parts = [
+          `A user from table '${tableName}' is missing one or more of the follow properties: uid, username, ns`,
+          `User = ${itemStr}.`,
+        ];
+        const suggestions = [
+          `Check the table '${tableName}' and populate the missing properties for the user and then retry the post deployment.`,
+        ];
+        const message = parts.join('\n');
+        tableReport.addFindings(item, [{ status: 'missingUidOrUsernameOrNs' }]);
+        throw new ErrorWithSuggestions(message, suggestions);
+      }
+
+      const objId = JSON.stringify({ ns, username });
+      this.uidMap[uid] = objId;
+      this.objIdMap[objId] = uid;
+      this.items[uid] = item;
+    });
+  }
+
+  // Given objId (either as a json string or json object { ns, username }), find the uid
+  uidLookup(objId) {
+    const id = _.isObject(objId) ? JSON.stringify(objId) : objId;
+    return this.objIdMap[id];
+  }
+
+  // Takes the old users items, if they have a matching new user items, they are left alone.
+  // Otherwise a new user item is created (in memory) and added to the internal items list.
+  // Any newly created users items are returned.
+  //
+  // The old user items = an array of items. An item is the raw Item object returned by DynamoDB
+  // which is a map of AttributeValues.
+  // Example: item = { username: { 'S': 'root' }, ns: { 'S': 'internal' }, ... }
+  // For more information about what an AttributeValue is, see
+  // https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_AttributeValue.html
+  async merge(oldUsersItems) {
```

Comment:
merge seems to be a bad method name. Does this method actually save anything to Dynamo or just provides a list of things to save. If it provides a list perhaps `getNewUserItems` is a better name?

Reply:
fixed